### PR TITLE
feat: hnswlib document index

### PR DESCRIPTION
**Pull request by JohannesMessner**
Goals:

First implementation of a Document Store.

**Design doc**: https://lightning-scent-57a.notion.site/Document-Stores-v2-design-doc-f11d6fe6ecee43f49ef88e0f1bf80b7f

**Usage Example**:

```python
# create a Document Store
from docarray.storage.backends.HnswDocStore import HnswDocumentIndex
from pydantic import Field
from docarray.typing import NdArray
from docarray import DocumentArray, BaseDocument

class MyDocument(BaseDocument):
    tens: NdArray[10]

store = HnswDocumentIndex[MyDocument](work_dir='path/to/work/dir')

# index some data    
data = DocumentArray[MyDocument]([MyDocument(tens=np.random.randn(10)) for i in range(10)])
store.index(data)

# find by np array
result = store.find(np.random.random((10,)), embedding_field='tens', limit=5)
result = store.find_batched(np.random.random((2, 10)), embedding_field='tens', limit=5)

# find by Document
result = store.find(MyDocument(tens=np.random.randn(10)), search_field='tens', limit=5)
result = store.find_batched(DocumentArray[MyDocument]([MyDocument(tens=np.random.randn(10)) for i in range(10)]), search_field='tens', limit=5)

# delete data
del store[data.id]  # delete all

# advanced configs
class MyDocument(BaseDocument):
    tens_one: NdArray = Field(dim=10, space='cosine')
    tens_two: NdArray = Field(dim=10, space='l2')
store = HnswDocumentStore[MyDocument](work_dir='path/to/work/dir')

# select search field
data = DocumentArray[MyDocument]([MyDocument(tens_one=np.random.randn(10), tens_two=np.random.randn(10)) for i in range(10)])
store.index(data)
result = store.find(np.random.random((10,)), search_field='tens_one', limit=5)
result = store.find(np.random.random((10,)), search_field='tens_two', limit=5)

# use query builder
q = store.build_query()
q.find(query=data[0], search_field='tens_one', limit=5)
q.find(query=data[1], search_field='tens_two', limit=5)
q.filter(filter_query={'tens_two': {'$exists': True}})
result = store.execute_query(q.build())

# nested search
class InnerDocument(BaseDocument):
    tens_one: NdArray = Field(dim=10)

class MyDocument(BaseDocument):
    d: InnerDocument
    tens: NdArray = Field(dim=10)

store = HnswDocumentStore[MyDocument](work_dir='path/to/work/dir')
data = DocumentArray[MyDocument]([MyDocument(d=InnerDocument(tens_one=np.random.randn(10)), tens=np.random.randn(10)) for i in range(10)])
store.index(data)

result = store.find(np.random.random((10,)), search_field='tens', limit=5)
result = store.find(np.random.random((10,)), search_field='d__tens_one', limit=5)

```

**ToDo:**
- [x] basic search
- [x] automatic conversions in the base class
- [x] query builder interface
  - [x] autocomplete for method kwargs
- [x] think about other interface methods such as `num_docs()` etc
- [x] think about abstraction layers
- [x] infer tensor dimensionality from type hint
- [x] handle config passing properly
- [x] persistence using SQLite
  - [ ] proper error messages when things go wrong
- [x] handle HNSWLib nuances around saving/loading and changing configurations
- [x] deleting documents
~~- [ ] nested access syntax: change `__` to `.`~~
- [ ] test torch and tf
- [ ] enable arbitrary data types in fields
- [ ] subindices
- [ ] tests

**Note:** unchecked boxes above have been moved to separate PR

### CODE REVIEW

**Review by JoanFM — COMMENTED** [review]
Let's call `DocumentIndexers` and not `stores`

File: docarray/storage/backends/HnswDocStore.py
Comment (JoanFM) [inline docarray/storage/backends/HnswDocStore.py:83]: it is not clear to me how and where does people provide this.

**Review by JoanFM — COMMENTED** [review]
It is not clear to me how composability of the QueryBuilder work

File: docarray/storage/backends/HnswDocStore.py
Comment (JohannesMessner) [reply docarray/storage/backends/HnswDocStore.py:83]: what do you mean by "people"?
Backend devs put the dataclass as inner class in their implementation, and users can pass it to the `configure` method

**Comment by JohannesMessner** [comment]
> It is not clear to me how composability of the QueryBuilder work

I think we need to leave that up to the individual backends. Since every backend defines their queries differently, I don't think we can compose queries for them. So the base query builder just collects all the calls and args.

Actually I would love to automate this more across different backends, but couldn't think of an approach. Do you have an idea?

File: docarray/storage/abstract_doc_store.py
Comment (JoanFM) [inline docarray/storage/abstract_doc_store.py]:
```suggestion
        self._db_config = db_config or self._db_config_cls(**kwargs)
```

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by JohannesMessner** [comment]
I would now call this PR *ready to review & merge*.
**Note**: *Only* the part considering the `AbstractDocumentStore` is considered ~done, there is still more work to do on the HNSWLib implementation. But this will be tackled in a separate PR.
That way we can unblock work on other backends/DBs.

So when reviewing please focus on:
- `AbstractDocumentStore` implementation
- how the HNSWLib implementation interacts with the abstraction layers

File: docarray/storage/abstract_doc_store.py
Comment (samsja) [inline docarray/storage/abstract_doc_store.py:547]: here you need to do the same magic as in DOcumentArray to handle the case where the item is a string or a typevar

Comment (samsja) [reply docarray/storage/abstract_doc_store.py:547]: maybe at some point we might want to abstract aways this templating behavior. But not now I guess

**Review by JoanFM — CHANGES_REQUESTED** [review]
can we please change the folder names to something related to indices?

**Review by JoanFM — CHANGES_REQUESTED** [review]
refactor storage to index in name of folders, files, packages, tests, etc ...

File: docarray/storage/abstract_doc_store.py
Comment (JohannesMessner) [reply docarray/storage/abstract_doc_store.py:547]: you're right

**Review by JoanFM — CHANGES_REQUESTED** [review]
the `folder` storage should be renamed

**Review by JoanFM — APPROVED** [review]

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (samsja) [inline docarray/doc_index/backends/hnswlib_doc_index.py:215]: as I said before, I think this one should accept `kwargs` but actually be specific

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (samsja) [inline docarray/doc_index/backends/hnswlib_doc_index.py:55]: why `**kwargs` here ? could we do smth so that the docstring are updatrd accordingly ? (could do in another PR but would need a ticket)

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (samsja) [inline docarray/doc_index/backends/hnswlib_doc_index.py:141]: I am not sure to understand why we need `**kwargs` here. Is it only for compatibility with the mother class ? Is this only a mypy problem ? I would rather silence mypy than offuscated value under a `**kwargs`

**Review by samsja — CHANGES_REQUESTED** [review]
LGTM. I just worry about the utilization of `**kwargs` I think we should never allow them

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JoanFM) [reply docarray/doc_index/backends/hnswlib_doc_index.py:141]: I believe is not a bad strategy to make API compatible with other Indexes

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:215]: After thinking about it, I don't think it works. If this doesn't take `**kwargs` but the public `find` does (as it needs to), then how does `find` pass the **kwargs to `_find`?
At runtime the **kwargs will be passed to a specific `_find` implementation that accepts all the needed kwargs, but as far as mypy is concerned, they are passed to the abstract `_find`.

So if that doesn't take `**kwargs` it will not work.

(Nor should it, from a typing perspective, bc we should not rely on the fact that a subclass implementation does something that the base class implementation doesn't do)

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:55]: It is convenience for the user.
Instead of creating a `config` object they can pass all the config params individually, as **kwargs, and it will just work.

Otherwise the user would have to import the config class (differently for each backend), instantiate it with some settings, and pass it, which is cumbersome.

But it can also be useful in certain settings, e.g. if a user want to clearly define all config setting globally. So both options exist.

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:141]: same logic as with `_find`. No value if obfuscated in the **kwargs, the next line checks that **kwargs is empty.

Comment (samsja) [reply docarray/doc_index/backends/hnswlib_doc_index.py:141]: yes but I think this can be avoided

Comment (samsja) [reply docarray/doc_index/backends/hnswlib_doc_index.py:141]: > I believe is not a bad strategy to make API compatible with other Indexes

I think we can achieve this without `**kwargs`

**Comment by JohannesMessner** [comment]
I added an .md with instructions for creating a new Doc Index class, I'll address the **kwargs comments tomorrow

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (AnneYang720) [inline docarray/doc_index/backends/hnswlib_doc_index.py:57]: can we move this line to `super().__init__` as well? like
`self._db_config = cast(self.DBConfig, self._db_config)`

File: docs/tutorials/add_doc_index.md
Comment (AnneYang720) [inline docs/tutorials/add_doc_index.md:89]:
```suggestion
You can declare allowed fields and default values for your `_runtime_config `, but you will see that later.
```

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (AnneYang720) [inline docarray/doc_index/backends/hnswlib_doc_index.py:47]: what about basic python type like `str`, `int`...

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (AnneYang720) [inline docarray/doc_index/backends/hnswlib_doc_index.py:130]: return `np.ndarray` or `allowed_type `

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:57]: No I don't think so. The issue is that mypy wants `self._db_config` to be of type `HnswDocumentIndex.DBConfig`, and not of type `AbstractDocumentIndex.DBConfig`, and the abstract class (`super()`) does not know about `HnswDocumentIndex.DBConfig`.
But luckily this is only about mypy, the code will run perfectly fine also without this line

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:47]: I still need to add support for those, but will do that in a separate PR.

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:130]: I don't think that implements the same logic

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (AnneYang720) [inline docarray/doc_index/backends/hnswlib_doc_index.py:108]: do we need to keep this `dim` here when class `_Column` has `n_dim`?

File: docs/tutorials/add_doc_index.md
Comment (samsja) [inline docs/tutorials/add_doc_index.md:1]: can you add it in the documentation ?

Comment (JohannesMessner) [reply docs/tutorials/add_doc_index.md:1]: ok

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:108]: good point, let me double check this

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (AnneYang720) [inline docarray/doc_index/backends/hnswlib_doc_index.py:307]: can we create a `hnswlib.Index` with no params instead if it will load content from disk on next line

Comment (AnneYang720) [reply docarray/doc_index/backends/hnswlib_doc_index.py:307]: I just read hnswlib documentation. no problem here

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:108]: Ok to clarify: The `n_dim` in the `_Column` is taken from the type parameter, e.g. `NdArray[512]`, then `n_dim` will be 512. The `dim` here is just a parameter that people can pass to `Field()`. So in the `_Column` it could be that `n_dim` is empty while `dim` isn't, or vice versa, or both are empty, etc.

File: docarray/doc_index/backends/hnswlib_doc_index.py
Comment (JohannesMessner) [reply docarray/doc_index/backends/hnswlib_doc_index.py:307]: I'll keep it in mind for the follow-up PR that refines the HNSWlib implementation :)

**Review by samsja — APPROVED** [review]

**Review by samsja — CHANGES_REQUESTED** [review]
can we mark all of the test as "slow" and "docstore" related ?

**Comment by JohannesMessner** [comment]
> can we mark all of the test as "slow" and "docstore" related ?

yep makes sense

File: tests/doc_index/hnswlib/test_index_get_del.py
Comment (samsja) [inline tests/doc_index/hnswlib/test_index_get_del.py:9]: oh did not know you could do this interesting

**Review by samsja — APPROVED** [review]

**Pull request merged**
